File: java-module/src/test/java/jvm/language/compare/java/cards/DeckOfCardsTest.java
package jvm.language.compare.java.cards;

import java.util.EnumSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

import org.eclipse.collections.api.multimap.list.ListMultimap;
import org.eclipse.collections.api.set.MutableSet;
import org.eclipse.collections.impl.test.Verify;
import org.junit.Assert;
import org.junit.Test;

public class DeckOfCardsTest
{
    private JavaDeckOfCards jdkDeck = new JavaDeckOfCards();
    private JavaStreamDeckOfCards streamsDeck = new JavaStreamDeckOfCards();

    @Test
    public void allCards()
    {
        Verify.assertSize(52, this.jdkDeck.getCards());
        Assert.assertEquals(this.jdkDeck.getCards(), this.streamsDeck.getCards());
    }

    @Test
    public void diamonds()
    {
        Verify.assertSize(13, this.jdkDeck.diamonds());
        Verify.assertAllSatisfy(this.jdkDeck.diamonds(), Card::isDiamonds);
        Assert.assertEquals(this.jdkDeck.diamonds(), this.streamsDeck.diamonds());
    }

    @Test
    public void hearts()
    {
        Verify.assertSize(13, this.jdkDeck.hearts());
        Verify.assertAllSatisfy(this.jdkDeck.hearts(), Card::isHearts);
        Assert.assertEquals(this.jdkDeck.hearts(), this.streamsDeck.hearts());
    }

    @Test
    public void spades()
    {
        Verify.assertSize(13, this.jdkDeck.spades());
        Verify.assertAllSatisfy(this.jdkDeck.spades(), Card::isSpades);
        Assert.assertEquals(this.jdkDeck.spades(), this.streamsDeck.spades());
    }

    @Test
    public void clubs()
    {
        Verify.assertSize(13, this.jdkDeck.clubs());
        Verify.assertAllSatisfy(this.jdkDeck.clubs(), Card::isClubs);
        Assert.assertEquals(this.jdkDeck.clubs(), this.streamsDeck.clubs());
    }

    @Test
    public void deal()
    {
        this.jdkDeck.shuffle(new Random(1));
        this.streamsDeck.shuffle(new Random(1));

        MutableSet<Card> jdkHand = this.jdkDeck.deal(5);
        Set<Card> streamsHand = this.streamsDeck.deal(5);

        Verify.assertSize(5, jdkHand);
        Assert.assertEquals(jdkHand, streamsHand);
    }

    @Test
    public void shuffleAndDealHands()
    {
        List<Set<Card>> jdkHands = this.jdkDeck.shuffleAndDeal(new Random(1), 5, 5);
        List<Set<Card>> streamsHands = this.streamsDeck.shuffleAndDeal(new Random(1), 5, 5);
        Verify.assertSize(5, jdkHands);
        Verify.assertAllSatisfy(jdkHands, each -> each.size() == 5);
        Assert.assertEquals(jdkHands, streamsHands);
    }

    @Test
    public void dealHands()
    {
        this.jdkDeck.shuffle(new Random(1));
        this.streamsDeck.shuffle(new Random(1));

        List<Set<Card>> jdkHands = this.jdkDeck.dealHands(5, 5);
        List<Set<Card>> streamsHands = this.streamsDeck.dealHands(5, 5);

        Verify.assertSize(5, jdkHands);
        Verify.assertAllSatisfy(jdkHands, each -> each.size() == 5);
        Assert.assertEquals(jdkHands, streamsHands);
    }

    @Test
    public void cardsBySuit()
    {
        ListMultimap<Suit, Card> jdkCardsBySuit = this.jdkDeck.getCardsBySuit();
        Map<Suit, List<Card>> streamsCardsBySuit = this.streamsDeck.getCardsBySuit();

        Assert.assertEquals(jdkCardsBySuit.get(Suit.CLUBS), streamsCardsBySuit.get(Suit.CLUBS));
        Assert.assertEquals(jdkCardsBySuit.get(Suit.DIAMONDS), streamsCardsBySuit.get(Suit.DIAMONDS));
        Assert.assertEquals(jdkCardsBySuit.get(Suit.HEARTS), streamsCardsBySuit.get(Suit.HEARTS));
        Assert.assertEquals(jdkCardsBySuit.get(Suit.SPADES), streamsCardsBySuit.get(Suit.SPADES));
    }

    @Test
    public void countsBySuit()
    {
        Verify.assertAllSatisfy(EnumSet.allOf(Suit.class), suit -> this.jdkDeck.countsBySuit().occurrencesOf(suit) == 13);
        Verify.assertAllSatisfy(EnumSet.allOf(Suit.class), suit -> this.streamsDeck.countsBySuit().get(suit) == 13);
    }

    @Test
    public void countsByRank()
    {
        Verify.assertAllSatisfy(EnumSet.allOf(Rank.class), rank -> this.jdkDeck.countsByRank().occurrencesOf(rank) == 4);
        Verify.assertAllSatisfy(EnumSet.allOf(Rank.class), rank -> this.streamsDeck.countsByRank().get(rank) == 4);
    }

    @Test
    public void dealOneCard()
    {
        this.jdkDeck.shuffle(new Random(1));
        Card card1jdk = this.jdkDeck.dealOneCard();
        Assert.assertEquals(51, this.jdkDeck.cardsLeftInDeck());
        Card card2jdk = this.jdkDeck.dealOneCard();
        Assert.assertEquals(50, this.jdkDeck.cardsLeftInDeck());
        Assert.assertNotEquals(card1jdk, card2jdk);

        this.streamsDeck.shuffle(new Random(1));
        Card card1streams = this.streamsDeck.dealOneCard();
        Assert.assertEquals(51, this.streamsDeck.cardsLeftInDeck());
        Card card2streams = this.streamsDeck.dealOneCard();
        Assert.assertEquals(50, this.streamsDeck.cardsLeftInDeck());
        Assert.assertNotEquals(card1streams, card2streams);
    }
}

File: kotlin-module/src/test/kotlin/jvm/language/compare/kotlin/cards/DeckOfCardsTest.kt
package jvm.language.compare.kotlin.cards

import jvm.language.compare.java.cards.JavaDeckOfCards
import org.eclipse.collections.api.list.ListIterable
import org.eclipse.collections.impl.set.mutable.SetAdapter
import org.eclipse.collections.impl.test.Verify
import org.junit.Assert
import org.junit.Test

class DeckOfCardsTest {
    private val kotlinDeck = KotlinDeckOfCards()
    private val jdkDeck = JavaDeckOfCards()

    @Test
    fun allCards() {
        Verify.assertSize(52, this.kotlinDeck.cards)
        Verify.assertSize(52, HashSet(this.kotlinDeck.cards))
        javaKotlinCardEquals(this.jdkDeck.cards, this.kotlinDeck.cards)
    }

    @Test
    fun diamonds() {
        Verify.assertSize(13, this.kotlinDeck.diamonds())
        Verify.assertAllSatisfy(this.kotlinDeck.diamonds(), Card::isDiamonds)
        javaKotlinCardEquals(this.jdkDeck.diamonds(), this.kotlinDeck.diamonds())
    }

    @Test
    fun hearts() {
        Verify.assertSize(13, this.kotlinDeck.hearts())
        Verify.assertAllSatisfy(this.kotlinDeck.hearts(), Card::isHearts)
        javaKotlinCardEquals(this.jdkDeck.hearts(), this.kotlinDeck.hearts())
    }

    @Test
    fun spades() {
        Verify.assertSize(13, this.kotlinDeck.spades())
        Verify.assertAllSatisfy(this.kotlinDeck.spades(), Card::isSpades)
        javaKotlinCardEquals(this.jdkDeck.spades(), this.kotlinDeck.spades())
    }

    @Test
    fun clubs() {
        Verify.assertSize(13, this.kotlinDeck.clubs())
        Verify.assertAllSatisfy(this.kotlinDeck.clubs(), Card::isClubs)
        javaKotlinCardEquals(this.jdkDeck.clubs(), this.kotlinDeck.clubs())
    }

    @Test
    fun deal() {
        this.kotlinDeck.shuffle(java.util.Random(1))
        this.jdkDeck.shuffle(java.util.Random(1))

        val kotlinHand = this.kotlinDeck.deal(5)
        val jdkHand = this.jdkDeck.deal(5)

        Verify.assertSize(5, kotlinHand)
        javaKotlinCardEquals(jdkHand, kotlinHand)
    }

    @Test
    fun shuffleAndDealHands() {
        val kotlinHands = this.kotlinDeck.shuffleAndDeal(java.util.Random(1), 5, 5)
        val jdkHands = this.jdkDeck.shuffleAndDeal(java.util.Random(1), 5, 5)
        Verify.assertSize(5, kotlinHands)
        Assert.assertEquals(kotlinHands.size, jdkHands.size)
        kotlinHands.forEach { kotlinHand -> Verify.assertSize(5, kotlinHand) }


        kotlinHands.forEachIndexed { index, kotlinHand ->
            var jdkHand = SetAdapter.adapt(jdkHands.get(index))
            javaKotlinCardEquals(jdkHand, kotlinHand)
        }
    }

    @Test
    fun dealHands() {
        this.kotlinDeck.shuffle(java.util.Random(1))
        this.jdkDeck.shuffle(java.util.Random(1))

        val kotlinHands = this.kotlinDeck.dealHands(5, 5)
        val jdkHands = this.jdkDeck.dealHands(5, 5)

        Verify.assertSize(5, kotlinHands)
        Assert.assertEquals(kotlinHands.size, jdkHands.size)
        kotlinHands.forEach { kotlinHand -> Verify.assertSize(5, kotlinHand) }

        kotlinHands.forEachIndexed { index, kotlinHand ->
            var jdkHand = SetAdapter.adapt(jdkHands.get(index))
            javaKotlinCardEquals(jdkHand, kotlinHand)
        }
    }

    @Test
    fun cardsBySuit() {
        val kotlinCardsBySuit = this.kotlinDeck.cardsBySuit
        val jdkCardsBySuit = this.jdkDeck.cardsBySuit

        javaKotlinCardEquals(
                jdkCardsBySuit.get(jvm.language.compare.java.cards.Suit.CLUBS),
                kotlinCardsBySuit.get(Suit.CLUBS))
        javaKotlinCardEquals(
                jdkCardsBySuit.get(jvm.language.compare.java.cards.Suit.DIAMONDS),
                kotlinCardsBySuit.get(Suit.DIAMONDS))
        javaKotlinCardEquals(
                jdkCardsBySuit.get(jvm.language.compare.java.cards.Suit.HEARTS),
                kotlinCardsBySuit.get(Suit.HEARTS))
        javaKotlinCardEquals(
                jdkCardsBySuit.get(jvm.language.compare.java.cards.Suit.SPADES),
                kotlinCardsBySuit.get(Suit.SPADES))
    }

    @Test
    fun countsBySuit() {
        Verify.assertAllSatisfy(Suit.values().asIterable()) { suit -> this.kotlinDeck.countsBySuit().get(suit) == 13 }
        Verify.assertAllSatisfy(jvm.language.compare.java.cards.Suit.values().asIterable()) { suit -> this.jdkDeck.countsBySuit().occurrencesOf(suit) == 13 }
    }

    @Test
    fun countsByRank() {
        Verify.assertAllSatisfy(Rank.values().asIterable()) { rank -> this.kotlinDeck.countsByRank().get(rank) == 4 }
        Verify.assertAllSatisfy(jvm.language.compare.java.cards.Rank.values().asIterable()) { rank -> this.jdkDeck.countsByRank().occurrencesOf(rank) == 4 }
    }

    @Test
    fun dealOneCard() {
        this.kotlinDeck.shuffle(java.util.Random(1))
        var card1kotlin = this.kotlinDeck.dealOneCard()
        Assert.assertEquals(51, this.kotlinDeck.cardsLeftInDeck())
        var card2kotlin = this.kotlinDeck.dealOneCard()
        Assert.assertEquals(50, this.kotlinDeck.cardsLeftInDeck())
        Assert.assertNotEquals(card1kotlin, card2kotlin)

        this.jdkDeck.shuffle(java.util.Random(1))
        var card1jdk = this.jdkDeck.dealOneCard()
        Assert.assertEquals(51, this.jdkDeck.cardsLeftInDeck())
        var card2jdk = this.jdkDeck.dealOneCard()
        Assert.assertEquals(50, this.jdkDeck.cardsLeftInDeck())
        Assert.assertNotEquals(card1jdk, card2jdk)

        javaKotlinCardEquals(this.jdkDeck.cards, this.kotlinDeck.cards)
    }

    companion object {
        fun javaKotlinCardEquals(javaCards: ListIterable<jvm.language.compare.java.cards.Card>,
                                 kotlinCards: Iterable<Card>?) {
            Assert.assertEquals(javaCards.toString(), kotlinCards.toString())
        }

        fun javaKotlinCardEquals(javaCards: org.eclipse.collections.api.set.MutableSet<jvm.language.compare.java.cards.Card>,
                                 kotlinCards: MutableSet<Card>) {
            Assert.assertEquals(javaCards.size, kotlinCards.size)

            // Must collect to Set as order is not guaranteed so, the toString() should be called on each element.
            var javaSet = javaCards.collect { card -> card.toString() }
            var kotlinSet = kotlinCards.map { card -> card.toString() }.toSet()
            Assert.assertEquals(javaSet, kotlinSet)
        }
    }
}
